tsl/src/remote/dist_copy.c

```diff
@@ -254,19 +252,36 @@ static void
 start_remote_copy_on_new_connection(CopyConnectionState *state, TSConnection *connection)
 {
 	PGconn *pg_conn = remote_connection_get_pg_conn(connection);
+
 	if (PQisnonblocking(pg_conn))
 		ereport(ERROR,
 				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
 				 errmsg("distributed copy doesn't support non-blocking connections")));
 
 	if (!list_member_ptr(state->connections_in_use, connection))
 	{
-		PGresult *res = PQexec(pg_conn, state->outgoing_copy_cmd);
+		PGresult *volatile res = NULL;
 
-		if (PQresultStatus(res) != PGRES_COPY_IN)
-			ereport(ERROR,
-					(errcode(ERRCODE_CONNECTION_FAILURE),
-					 errmsg("unable to start remote COPY on data node (%d)", PQresultStatus(res))));
+		PG_TRY();
+		{
```

Comment:
No, this is to ensure that the PGresult is always released, even if we encounter an error or throw an error ourselves. The PGresult is malloced and will cause a permanent memory leak if we don't guarantee its release. And the only safe way to do that is a try-catch.

Reply:
But if there is an error in `PQexec` nothing will be assigned to `res` (because the error function will do a `longjmp` to the exception handler directly) so then you do not need any cleanup code. The only place where `res` can have a value and an error can be thrown is from the `ereport` below, and the only reason you cannot call `PQclear` before calling `ereport` is because `PQresultErrorMessage` references the internals of `res`, so removing the entire try-catch and doing this instead is easier to understand:

```
if (PQresultStatus(res) != PGRES_COPY_IN)
{
  char *msg = pchomp(PQresultErrorMessage(res));
  PQclear(res);
				ereport(ERROR,
						(errcode(ERRCODE_CONNECTION_FAILURE),
						 errmsg("unable to start remote COPY on data node"),
						 errdetail("Remote command error: %s", msg)));
}
```

Note that you only need to copy the error message, because that's all you use.

Reply:
I am not sure I understand what the improvement is apart from "just different".

FWIW, this is the standard pattern for handling PGresult, e.g., in the `postgres_fdw`. And, it is safe to future changes, e.g., if someone adds a function call  that can throw and error between the PQexec and the error checking.

Reply:
Easier to follow the logic, less duplicated code, that's all. This is a matter of taste. I have approved the PR, feel free to push.